clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy

```diff
@@ -319,7 +317,7 @@ class ReservationReportCachingAgent implements CachingAgent, CustomScheduledAgen
             def osType = operatingSystemType(it.productDescription)
             def reservation = getReservation(region.name, it.availabilityZone, osType.name, it.instanceType)
             reservation.totalReserved.addAndGet(it.instanceCount)
-
+            def vpcOnlyAccounts = determineVpcOnlyAccounts()
```

Comment:
We can't really run this in a tight loop since it makes a describe call to ec2 and is going to get throttled (it certainly 

I played around locally with:
`final AtomicReference<Set<String>> vpcOnlyAccountsReference = new AtomicReference<>()`

and 

```
  private Set<String> determineVpcOnlyAccounts() {
    if (vpcOnlyAccountsReference.get() != null) {
      return vpcOnlyAccountsReference.get();
    }

    def vpcOnlyAccounts = new HashSet<>()

    accounts.each { credentials ->
      def amazonEC2 = amazonClientProvider.getAmazonEC2(credentials, credentials.regions[0].name)
      def describeAccountAttributesResult = amazonEC2.describeAccountAttributes(
        new DescribeAccountAttributesRequest().withAttributeNames("supported-platforms")
      )
      if (describeAccountAttributesResult.accountAttributes[0].attributeValues*.attributeValue == ["VPC"]) {
        vpcOnlyAccounts << credentials.name
      }
    }

    log.info("VPC Only Accounts: ${vpcOnlyAccounts.join(", ")}")
    vpcOnlyAccountsReference.set(vpcOnlyAccounts)

    return vpcOnlyAccounts
  }
```

and

```    def vpcOnlyAccounts = determineVpcOnlyAccounts()

    credentials.regions.each { AmazonCredentials.AWSRegion region ->
        log.info("Fetching reservation report for ${credentials.name}:${region.name}")
        long startTime = System.currentTimeMillis()
```

Reply:
One of aims of the account registration RFC is to allow for dynamic account loading, so I think it makes sense for `vpcOnlyAccounts` to be updated to reflect account changes. Would something like this work? I will also need to update `AmazonCredentialsLifecycleHandler` to update these fields in case of credentials deletion and updates.

```
final Set<String> vpcOnlyAccounts = ConcurrentHashMap.newKeySet()
final Set<String> checkedAccounts = ConcurrentHashMap.newKeySet()
```
```
private synchronized Set<String> determineVpcOnlyAccounts() {
    accounts.each { credentials ->
      if (!checkedAccounts.contains(credentials.getName())) {
        def amazonEC2 = amazonClientProvider.getAmazonEC2(credentials, credentials.regions[0].name)
        def describeAccountAttributesResult = amazonEC2.describeAccountAttributes(
          new DescribeAccountAttributesRequest().withAttributeNames("supported-platforms")
        )
        if (describeAccountAttributesResult.accountAttributes[0].attributeValues*.attributeValue == ["VPC"]) {
          vpcOnlyAccounts.add(credentials.getName())
        }
        checkedAccounts.add(credentials.getName())
      }
    }
    log.info("VPC Only Accounts: ${vpcOnlyAccounts.join(", ")}")
    return vpcOnlyAccounts
  }
```

Reply:
Ah yes, of course.

Tbh I don't know if we even effectively support non-vpc accounts anymore. It's also unlikely that this particular caching agent is widely run.

I think even maintaining a Map<String, Boolean> for whether an account is vpc or not would suffice.

Reply:
I did roll out my earlier quick change here and it's looking better and not causing excessive rate limiting.

Reply:
I wonder if it makes sense to determine VPC only accounts in `AmazonCredentialsLifecycleHandler` and when these credentials get added rather than delaying it to be done inside an agent? 

The property is inherent to the accounts anyway and the extra call shouldn't introduce a crazy amount of overhead.

Reply:
I have moved `determineVpcOnlyAccounts()` functionality to `AmazonCredentialsLifecycleHandler`. It will now determine if a given account is VPC only or not when adding accounts. The account is removed from map when account is deleted.